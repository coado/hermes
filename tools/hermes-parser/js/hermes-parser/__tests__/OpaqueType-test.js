/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

import {
  printForSnapshotESTree,
  parseForSnapshotESTree,
  printForSnapshotBabel,
  parseForSnapshotBabel,
} from '../__test_utils__/parse';

describe('OpaqueType', () => {
  describe('Basic', () => {
    const code = `
        declare opaque type Foo super Bar extends Baz;
        opaque type Foo super Bar extends Baz = Boz;
    `;

    test('ESTree', async () => {
      expect(await parseForSnapshotESTree(code)).toMatchSnapshot();
      expect(await printForSnapshotESTree(code)).toBe(`declare opaque type Foo;
opaque type Foo = Boz;`);
    });

    test('Babel', async () => {
      expect(await parseForSnapshotBabel(code)).toMatchSnapshot();
      expect(await printForSnapshotBabel(code)).toMatchInlineSnapshot(`
        "declare opaque type Foo;
        opaque type Foo = Boz;"
      `);
    });
  });
});
