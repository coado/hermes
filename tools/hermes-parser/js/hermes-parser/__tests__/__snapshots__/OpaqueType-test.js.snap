// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpaqueType Basic Babel 1`] = `
{
  "body": [
    {
      "id": {
        "name": "Foo",
        "type": "Identifier",
      },
      "impltype": null,
      "supertype": null,
      "type": "DeclareOpaqueType",
      "typeParameters": null,
    },
    {
      "id": {
        "name": "Foo",
        "type": "Identifier",
      },
      "impltype": {
        "id": {
          "name": "Boz",
          "type": "Identifier",
        },
        "type": "GenericTypeAnnotation",
        "typeParameters": null,
      },
      "supertype": null,
      "type": "OpaqueType",
      "typeParameters": null,
    },
  ],
  "type": "Program",
}
`;

exports[`OpaqueType Basic ESTree 1`] = `
{
  "body": [
    {
      "id": {
        "name": "Foo",
        "optional": false,
        "type": "Identifier",
        "typeAnnotation": null,
      },
      "impltype": null,
      "lowerBound": {
        "id": {
          "name": "Bar",
          "optional": false,
          "type": "Identifier",
          "typeAnnotation": null,
        },
        "type": "GenericTypeAnnotation",
        "typeParameters": null,
      },
      "supertype": null,
      "type": "DeclareOpaqueType",
      "typeParameters": null,
      "upperBound": {
        "id": {
          "name": "Baz",
          "optional": false,
          "type": "Identifier",
          "typeAnnotation": null,
        },
        "type": "GenericTypeAnnotation",
        "typeParameters": null,
      },
    },
    {
      "id": {
        "name": "Foo",
        "optional": false,
        "type": "Identifier",
        "typeAnnotation": null,
      },
      "impltype": {
        "id": {
          "name": "Boz",
          "optional": false,
          "type": "Identifier",
          "typeAnnotation": null,
        },
        "type": "GenericTypeAnnotation",
        "typeParameters": null,
      },
      "lowerBound": {
        "id": {
          "name": "Bar",
          "optional": false,
          "type": "Identifier",
          "typeAnnotation": null,
        },
        "type": "GenericTypeAnnotation",
        "typeParameters": null,
      },
      "supertype": null,
      "type": "OpaqueType",
      "typeParameters": null,
      "upperBound": {
        "id": {
          "name": "Baz",
          "optional": false,
          "type": "Identifier",
          "typeAnnotation": null,
        },
        "type": "GenericTypeAnnotation",
        "typeParameters": null,
      },
    },
  ],
  "type": "Program",
}
`;
